clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -0,0 +1,250 @@
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.PasswordCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.sasl.AuthorizeCallback;
+import javax.security.sasl.RealmCallback;
+import javax.security.sasl.Sasl;
+import javax.security.sasl.SaslClient;
+import javax.security.sasl.SaslException;
+
+import org.apache.kafka.common.network.Authenticator;
+import org.apache.kafka.common.network.NetworkSend;
+import org.apache.kafka.common.network.NetworkReceive;
+import org.apache.kafka.common.network.TransportLayer;
+import org.apache.kafka.common.security.auth.KafkaPrincipal;
+import org.apache.kafka.common.security.auth.PrincipalBuilder;
+import org.apache.kafka.common.security.kerberos.KerberosName;
+import org.apache.kafka.common.KafkaException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SaslClientAuthenticator implements Authenticator {
+
+    public enum SaslState {
+        INITIAL, INTERMEDIATE, COMPLETE, FAILED
+    }
+
+    private static final Logger LOG = LoggerFactory.getLogger(SaslClientAuthenticator.class);
+
+    private final Subject subject;
+    private final String servicePrincipal;
+    private final String host;
+    private final String node;
+
+    // assigned in `configure`
+    private SaslClient saslClient;
+    private String clientPrincipalName;
+    private TransportLayer transportLayer;
+
+    // buffers used in `authenticate`
+    private NetworkReceive netInBuffer;
+    private NetworkSend netOutBuffer;
+
+    private SaslState saslState = SaslState.INITIAL;
+
+    public SaslClientAuthenticator(String node, Subject subject, String servicePrincipal, String host) throws IOException {
+        this.node = node;
+        this.subject = subject;
+        this.host = host;
+        this.servicePrincipal = servicePrincipal;
+    }
+
+    public void configure(TransportLayer transportLayer, PrincipalBuilder principalBuilder, Map<String, ?> configs) throws KafkaException {
+        try {
+            this.transportLayer = transportLayer;
+
+            // determine client principal from subject.
+            Principal clientPrincipal = subject.getPrincipals().iterator().next();
+            this.clientPrincipalName = new KerberosName(clientPrincipal.getName()).toString();
+            this.saslClient = createSaslClient();
+        } catch (Exception e) {
+            throw new KafkaException("Failed to configure SaslClientAuthenticator", e);
+        }
+    }
+
+    private SaslClient createSaslClient() {
+        try {
+            return Subject.doAs(subject, new PrivilegedExceptionAction<SaslClient>() {
+                public SaslClient run() throws SaslException {
+                    LOG.debug("Client will use GSSAPI as SASL mechanism.");
+                    String[] mechs = {"GSSAPI"};
```

Comment:
Can the mechanism be made a configuration option? I haven't looked through the code yet to see if the implementation relies on this mechanism, but it will be good if it was configurable.